From f619ae89c97ec1f5b7564a82e66835e40dc2c59b Mon Sep 17 00:00:00 2001
From: "Li, Ning" <ning.li@intel.com>
Date: Wed, 30 Jan 2013 21:07:28 +0800
Subject: [PATCH 066/429] rtc: enable cmos rtc driver on Tangier

Tangier does not support vRTC. Use CMOS RTC as rtc driver.

Signed-off-by: Li, Ning <ning.li@intel.com>
---
 arch/x86/kernel/rtc.c | 32 ++++++++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/arch/x86/kernel/rtc.c b/arch/x86/kernel/rtc.c
index 198eb201ed3b..c487ad76e9fb 100644
--- a/arch/x86/kernel/rtc.c
+++ b/arch/x86/kernel/rtc.c
@@ -14,6 +14,7 @@
 #include <asm/time.h>
 #include <asm/mrst.h>
 #include <asm/rtc.h>
+#include <asm/io_apic.h>
 
 #ifdef CONFIG_X86_32
 /*
@@ -149,6 +150,26 @@ void read_persistent_clock(struct timespec *ts)
 	ts->tv_nsec = 0;
 }
 
+static int handle_mrfl_dev_ioapic(int irq)
+{
+	int ret = 0;
+	int ioapic;
+	struct io_apic_irq_attr irq_attr;
+
+	ioapic = mp_find_ioapic(irq);
+	if (ioapic >= 0) {
+		irq_attr.ioapic = ioapic;
+		irq_attr.ioapic_pin = irq;
+		irq_attr.trigger = 1;
+		irq_attr.polarity = 0; /* Active high */
+		io_apic_set_pci_routing(NULL, irq, &irq_attr);
+	} else {
+		pr_warn("can not find interrupt %d in ioapic\n", irq);
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
 
 static struct resource rtc_resources[] = {
 	[0] = {
@@ -172,6 +193,8 @@ static struct platform_device rtc_device = {
 
 static __init int add_rtc_cmos(void)
 {
+	int ret;
+
 #ifdef CONFIG_PNP
 	static const char * const  const ids[] __initconst =
 	    { "PNP0b00", "PNP0b01", "PNP0b02", };
@@ -193,8 +216,17 @@ static __init int add_rtc_cmos(void)
 
 	/* Intel MID platforms don't have ioport rtc */
 	if (mrst_identify_cpu())
+	/* Intel MID platforms don't have ioport rtc
+	 * except Tangier platform, which doesn't have vRTC
+	 */
+	if (intel_mid_identify_cpu() &&
+	    intel_mid_identify_cpu() != INTEL_MID_CPU_CHIP_TANGIER)
 		return -ENODEV;
 
+	ret = handle_mrfl_dev_ioapic(RTC_IRQ);
+	if (ret)
+		return ret;
+
 	platform_device_register(&rtc_device);
 	dev_info(&rtc_device.dev,
 		 "registered platform RTC device (no PNP device found)\n");
-- 
2.37.3

